File: ksheermitra/lib/models/delivery.dart
import 'product.dart';
import 'user.dart';

class Delivery {
  final String id;
  final String customerId;
  final String subscriptionId;
  final String productId;
  final String deliveryBoyId;
  final DateTime deliveryDate;
  final double quantity;
  final double amount;
  final String status;
  final String? notes;
  final User? customer;
  final Product? product;
  final User? deliveryBoy;
  final DateTime createdAt;
  final DateTime updatedAt;

  Delivery({
    required this.id,
    required this.customerId,
    required this.subscriptionId,
    required this.productId,
    required this.deliveryBoyId,
    required this.deliveryDate,
    required this.quantity,
    required this.amount,
    required this.status,
    this.notes,
    this.customer,
    this.product,
    this.deliveryBoy,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Delivery.fromJson(Map<String, dynamic> json) {
    return Delivery(
      id: json['id']?.toString() ?? '',
      customerId: json['customerId']?.toString() ?? '',
      subscriptionId: json['subscriptionId']?.toString() ?? '',
      productId: json['productId']?.toString() ?? '',
      deliveryBoyId: json['deliveryBoyId']?.toString() ?? '',
      deliveryDate: json['deliveryDate'] != null
          ? DateTime.tryParse(json['deliveryDate'].toString()) ?? DateTime.now()
          : DateTime.now(),
      quantity: json['quantity'] != null
          ? double.tryParse(json['quantity'].toString()) ?? 0.0
          : 0.0,
      amount: json['amount'] != null
          ? double.tryParse(json['amount'].toString()) ?? 0.0
          : 0.0,
      status: json['status']?.toString() ?? 'pending',
      notes: json['notes']?.toString(),
      customer: json['customer'] != null && json['customer'] is Map<String, dynamic>
          ? User.fromJson(json['customer'])
          : null,
      product: json['product'] != null && json['product'] is Map<String, dynamic>
          ? Product.fromJson(json['product'])
          : null,
      deliveryBoy: json['deliveryBoy'] != null && json['deliveryBoy'] is Map<String, dynamic>
          ? User.fromJson(json['deliveryBoy'])
          : null,
      createdAt: json['createdAt'] != null
          ? DateTime.tryParse(json['createdAt'].toString()) ?? DateTime.now()
          : DateTime.now(),
      updatedAt: json['updatedAt'] != null
          ? DateTime.tryParse(json['updatedAt'].toString()) ?? DateTime.now()
          : DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'customerId': customerId,
      'subscriptionId': subscriptionId,
      'productId': productId,
      'deliveryBoyId': deliveryBoyId,
      'deliveryDate': deliveryDate.toIso8601String(),
      'quantity': quantity,
      'amount': amount,
      'status': status,
      'notes': notes,
    };
  }

  bool get isPending => status == 'pending';
  bool get isDelivered => status == 'delivered';
  bool get isMissed => status == 'missed';
  bool get isCancelled => status == 'cancelled';
}

File: backend/src/models/Delivery.js
module.exports = (sequelize, DataTypes) => {
  const Delivery = sequelize.define('Delivery', {
    id: {
      type: DataTypes.UUID,
      defaultValue: DataTypes.UUIDV4,
      primaryKey: true
    },
    customerId: {
      type: DataTypes.UUID,
      allowNull: false,
      references: {
        model: 'Users',
        key: 'id'
      }
    },
    deliveryBoyId: {
      type: DataTypes.UUID,
      allowNull: false,
      references: {
        model: 'Users',
        key: 'id'
      }
    },
    subscriptionId: {
      type: DataTypes.UUID,
      allowNull: false,
      references: {
        model: 'Subscriptions',
        key: 'id'
      }
    },
    productId: {
      type: DataTypes.UUID,
      allowNull: true, // Changed to true for multi-product support
      references: {
        model: 'Products',
        key: 'id'
      }
    },
    deliveryDate: {
      type: DataTypes.DATEONLY,
      allowNull: false
    },
    quantity: {
      type: DataTypes.DECIMAL(10, 2),
      allowNull: true, // Changed to true for multi-product support
      validate: {
        min: 0,
        isDecimal: true
      }
    },
    status: {
      type: DataTypes.ENUM('pending', 'in-progress', 'delivered', 'missed', 'cancelled', 'failed'),
      allowNull: false,
      defaultValue: 'pending'
    },
    deliveredAt: {
      type: DataTypes.DATE,
      allowNull: true
    },
    amount: {
      type: DataTypes.DECIMAL(10, 2),
      allowNull: false,
      validate: {
        min: 0,
        isDecimal: true
      }
    },
    notes: {
      type: DataTypes.TEXT,
      allowNull: true
    },
    notificationSent: {
      type: DataTypes.BOOLEAN,
      defaultValue: false
    }
  }, {
    tableName: 'Deliveries',
    timestamps: true,
    paranoid: true,
    indexes: [
      {
        fields: ['customerId']
      },
      {
        fields: ['deliveryBoyId']
      },
      {
        fields: ['subscriptionId']
      },
      {
        fields: ['productId']
      },
      {
        fields: ['deliveryDate']
      },
      {
        fields: ['status']
      },
      {
        fields: ['deliveryDate', 'customerId']
      },
      {
        fields: ['deliveryDate', 'deliveryBoyId']
      }
    ]
  });

  Delivery.associate = (models) => {
    Delivery.belongsTo(models.User, {
      foreignKey: 'customerId',
      as: 'customer'
    });

    Delivery.belongsTo(models.User, {
      foreignKey: 'deliveryBoyId',
      as: 'deliveryBoy'
    });

    Delivery.belongsTo(models.Subscription, {
      foreignKey: 'subscriptionId',
      as: 'subscription'
    });

    Delivery.belongsTo(models.Product, {
      foreignKey: 'productId',
      as: 'product'
    });

    Delivery.hasMany(models.DeliveryItem, {
      foreignKey: 'deliveryId',
      as: 'items'
    });
  };

  return Delivery;
};

